libs/ml/include/ml/dataloaders/commodity_dataloader.hpp

```diff
@@ -0,0 +1,198 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include <fstream>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "core/random.hpp"
+#include "math/base_types.hpp"
+#include "math/tensor.hpp"
+#include "ml/dataloaders/dataloader.hpp"
+
+namespace fetch {
+namespace ml {
+
+namespace {
+std::pair<math::SizeType, math::SizeType> count_rows_cols(std::string filename,
+                                                          bool        transpose = false)
+{
+  // find number of rows and columns in the file
+  std::ifstream  file(filename);
+  std::string    buf;
+  std::string    delimiter = ",";
+  ulong          pos;
+  math::SizeType row{0};
+  math::SizeType col{0};
+  while (std::getline(file, buf, '\n'))
+  {
+    while (row == 0 && ((pos = buf.find(delimiter)) != std::string::npos))
+    {
+      buf.erase(0, pos + delimiter.length());
+      ++col;
+    }
+    ++row;
+  }
+  if (transpose)
+  {
+    return std::make_pair(col + 1, row);
+  }
+  else
+  {
+    return std::make_pair(row, col + 1);
+  }
+}
+}  // namespace
+
+template <typename LabelType, typename InputType>
+class CommodityDataLoader : DataLoader<LabelType, InputType>
+{
+
+public:
+  using DataType   = typename InputType::Type;
+  using ReturnType = std::pair<LabelType, std::vector<InputType>>;
+  using SizeType   = math::SizeType;
+
+  CommodityDataLoader(bool random_mode = false)
+    : DataLoader<LabelType, InputType>(random_mode)
+  {}
+
+  ~CommodityDataLoader() = default;
+
+  virtual ReturnType GetNext()
```

Comment:
please refactor this class so that the implementations are outside of the class body